Timeline

danielsteinkogler opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [x] I added a line to [changelog.md](/doc/changelog.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [x] References to issues, e.g. `close #X`, are in the commit messages and changelog.
- [x] The buildserver is happy.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
Otherwise please check these points when getting a PR done:
-->

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added automated tests or a [manual test protocol](../doc/tests/manual/protocol.md)
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)
- [ ] Exceptions to any guidelines are documented

## Review

<!--
Reviewers can copy&check the following to their review.
Also the checklist above can be used.
But also the PR creator should check these points when getting a PR done:
-->

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] I've read through the whole documentation
- [ ] I've checked conformity to guidelines

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
@badnames @absurd-turtle @4ydan @markus2330 
The PR contains a first version of the new timeline feature without event representation that can now be reviewed. I just set added and removed events to 0 for now and created some left over issues for known missing features.

@4ydan: currently there is no loading indicator implemented for the new timeline. That means that we currently don't have an visual indicator we can use in the e2e tests. I added a task to https://github.com/ElektraInitiative/PermaplanT/issues/1088 that we have to adapt e2e tests after implementing a loading indicator.

markus2330 reviewed (changes requested):
- [x] tour must be fixed, you cannot progress after Pflanzen-Ebene (7/16)
- [ ] status panel must be moved a bit up to avoid an overlap
- [ ] Firefox shows disturbing scroll bars (Chrome does not, the user experience is much more pleasant there)
- [ ] in storybook is still the old timeline for me and not the new http://localhost:6006/?path=/story/features-map-planning-components-timeline-timeline--default
- [ ] please mark the current day/month/year, with (PermaplanT) blue font (can be a follow up if it is not trivial)

4ydan reviewed (approved):
Good update of the e2e page object. 👍🏻

badnames commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.tsx:
```diff
@@ -0,0 +1,301 @@
+  );
+}
+
+function TimeLineDatePickerEventIndivatior({ added, removed }: { added: number; removed: number }) {
```
Did you mean indicator instead of indivatior?
```suggestion
function TimeLineDatePickerEventIndicator({ added, removed }: { added: number; removed: number }) {
```

badnames commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.tsx:
```diff
@@ -0,0 +1,301 @@
+  } = useGetTimeLineEvents();
+  const { i18n } = useTranslation();
+
+  const [selectedYearItem, setSelectedYearItem] = useState<YearItem>(
```
Is there a reason why you track this components state outside of our map store?

badnames commented on frontend/src/features/map_planning/utils/date-utils.ts:
```diff
@@ -16,3 +16,18 @@ export function convertToDateString(date: Date) {
+ * @returns short name of month in given language (e.g. 'Jan')
+ */
+
+export function getShortMonthNameFromNumber(monthNumber: number, language?: string) {
```
Consider adding a unit test for this function.

badnames reviewed (changes requested):
Overall good work. 
I like the design a lot, but there are still a few issues.

![image](https://github.com/ElektraInitiative/PermaplanT/assets/27735433/4da1c566-c453-4b3f-a1db-9ffece0670ad)

While testing your PR I noticed that the styling is broken on dark mode.

absurd-turtle commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+    if (value === undefined) return;
+
+    setSelectedItemKey(value);
+    scrollToIndex(items.findIndex((item) => item.key === value));
```
check how often scrollToIndex is called.
I think there might be some redundant calls when value is defined and the items change.

danielsteinkogler replied:
you are right, there a several unnecessary calls. i created a left over issue to optimize the the calls

markus2330 replied:
You can also link from such conversations to the issue. I assume you are referring to #1107

absurd-turtle reviewed (approved):
Code looks good to me! :)
Just check if the hooks are called more often than necessary.

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
jenkins build please

danielsteinkogler commented:
@markus2330 @badnames i fixed mentioned issues and added some tests. the code itself didn't change that much, but maybe you can test the behaviour again.

markus2330 commented:
jenkins build please

markus2330 commented on e2e/pages/maps/planting.py:
```diff
@@ -93,9 +97,30 @@ def check_plant_layer(self):
+                self._timeline_day_slider.press("ArrowLeft")
+
+    def change_map_date_by_months(self, delta_months: int):
+        """Changes the date by a given amount of months."""
```
Not used above in Scenarios?

markus2330 commented on e2e/pages/maps/planting.py:
```diff
@@ -41,7 +41,11 @@ def __init__(self, page: Page):
+        self._timeline_day_slider = page.get_by_test_id("timeline__day-slider")
+        self._timeline_month_slider = page.get_by_test_id("timeline__month-slider")
+        self._timeline_year_slider = page.get_by_test_id("timeline__year-slider")
         self._timeline_idle = page.get_by_test_id("timeline__date-form-idle")
```
Does this still exist?

markus2330 commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+  leftEndReached?: () => void;
+  /** Callback that fires when the right end of the slider is reached */
+  rightEndReached?: () => void;
+  /** initial value */
```
please describe the value

markus2330 commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+  };
+
+  const handleScroll = () => {
+    // If we are autiomatically scroll to an index we dont't want to handle the scroll event.
```
I don't know how to interpret the typos :wink:

markus2330 commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.test.tsx:
```diff
@@ -0,0 +1,64 @@
+    });
+
+    jest.runAllTimers();
+    expect(onSelectChange).toBeCalledWith('2020-11-01');
```
Test could be a bit more interesting, e.g. jump by 13 months or at least over the border of the year.

markus2330 commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.test.tsx:
```diff
@@ -0,0 +1,64 @@
+    });
+
+    jest.runAllTimers();
+    expect(onSelectChange).toBeCalledWith('2021-12-01');
```
Maybe here 100 years or so. <1930 should work (but can be also done in follow-up)

markus2330 commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.tsx:
```diff
@@ -0,0 +1,314 @@
+      triggerDateChangedInGuidedTour();
+    }, 500);
+    return () => clearTimeout(timeoutId);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
@Bushuo can you check if this is fine?

Bushuo replied:
If onSelectDate somehow changes, the effect would not trigger.
I have to look more into it. Maybe this is a valid usecase, to not trigger the effect.
Otherwise its fine

markus2330 commented on frontend/src/features/map_planning/utils/date-utils.test.ts:
```diff
@@ -0,0 +1,33 @@
+  });
+
+  it('should return correct english month names', () => {
+    expect(getShortMonthNameFromNumber(1, 'en-US')).toBe('Jan');
```
It should be the other way round: more trust that standard libs work and more exhaustive tests for own code :wink:

markus2330 reviewed (changes requested):
Great work! :heart:

Bushuo commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+
+  const detectAndNotifyIfBordersIsReached = () => {
+    const container = sliderContainerRef.current;
+    if (container) {
```
```tsx
if (container) {
```
to
```tsx
if (!container) return;
```
I generally prefer to invert the if conditions as guard clauses.
Makes for more readable code imo.
Especially if there are nested ifs inside.

Bushuo commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+
+      const threshold = 500;
+      if (scrollLeft < threshold) {
+        leftEndReached && leftEndReached();
```
```suggestion
        leftEndReached?.();
```

Bushuo commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+      if (scrollLeft < threshold) {
+        leftEndReached && leftEndReached();
+      } else if (scrollWidth - scrollLeft - containerWidth < threshold) {
+        rightEndReached && rightEndReached();
```
```suggestion
        rightEndReached?.();
```

Bushuo commented on frontend/src/features/map_planning/components/timeline/ItemSliderPicker.tsx:
```diff
@@ -0,0 +1,307 @@
+    }
+  };
+
+  useEffect(() => {
+    const container = sliderContainerRef.current;
+    container?.addEventListener('scroll', handleScroll);
+    container?.addEventListener('mousedown', handleMouseDown);
+    container?.addEventListener('wheel', handleMouseWheel);
+    document.addEventListener('mousemove', handleMouseMove);
+    document.addEventListener('mouseup', handleMouseUp);
+
+    return () => {
+      container?.removeEventListener('scroll', handleScroll);
+      container?.removeEventListener('mousedown', handleMouseDown);
+      container?.removeEventListener('wheel', handleMouseWheel);
+      document.removeEventListener('mousemove', handleMouseMove);
+      document.removeEventListener('mouseup', handleMouseUp);
+    };
+  });
```
Can this also be assigned in the JSX with `onScroll` etc?
React then handles the removal of the event listeners automatically.
Also this effect is called every render.

Bushuo commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.test.tsx:
```diff
@@ -0,0 +1,64 @@
+import { render } from '@testing-library/react';
+import ReactTestUtils, { act } from 'react-dom/test-utils';
+
+// ES6
```
```suggestion
```
Did you maybe forget to remove this comment?

Bushuo commented on frontend/src/features/map_planning/components/timeline/TimelineDatePicker.tsx:
```diff
@@ -0,0 +1,314 @@
+    }
+  };
+
+  function triggerDateChangedInGuidedTour(): void {
+    const changeDateEvent = new Event('dateChanged');
+    document.getElementById('timeline')?.dispatchEvent(changeDateEvent);
+  }
```
Please put this out of the component, because it is pure.

badnames reviewed (approved):
![grafik](https://github.com/ElektraInitiative/PermaplanT/assets/27735433/8cb5b78d-819e-47d4-89eb-4d5230580752)

Looks much better already.

danielsteinkogler commented:
Thanks for the great reviews :D

@markus2330 i added some new test cases and now also managed to check if the sliders itself are correctly synchronized using unit test.  i think we can add more after implementing the event indicators. I also added the loading indicator next to the current map date as we discussed together so that we can avoid flaky e2e tests.

@Bushuo i learned a lot while processing your suggestions 👍 I added some tasks with possible issues concerning react hooks here:
https://github.com/ElektraInitiative/PermaplanT/issues/1107.
I will ping you i if have questions when addressing this issues.
